test/extensions/filters/http/ext_authz/config_test.cc

```diff
@@ -3,95 +3,210 @@
-            return std::make_unique<NiceMock<Grpc::MockAsyncClient>>();
-          }));
-  cb(filter_callback);
-
-  Thread::ThreadPtr thread = Thread::threadFactoryForTest().createThread([&context, cb]() {
-    Http::MockFilterChainFactoryCallbacks filter_callback;
-    EXPECT_CALL(filter_callback, addStreamFilter(_));
-    // Execute the filter factory callback in another thread.
-    EXPECT_CALL(context.cluster_manager_.async_client_manager_,
-                getOrCreateRawAsyncClient(_, _, _, _))
-        .WillOnce(Invoke(
-            [](const envoy::config::core::v3::GrpcService&, Stats::Scope&, bool,
-               Grpc::CacheOption) { return std::make_unique<NiceMock<Grpc::MockAsyncClient>>(); }));
-    cb(filter_callback);
-  });
-  thread->join();
-}
 
-} // namespace
+class TestAsyncClientManagerImpl : public Grpc::AsyncClientManagerImpl {
+public:
+  TestAsyncClientManagerImpl(Upstream::ClusterManager& cm, ThreadLocal::Instance& tls,
+                             TimeSource& time_source, Api::Api& api,
+                             const Grpc::StatNames& stat_names)
+      : Grpc::AsyncClientManagerImpl(cm, tls, time_source, api, stat_names) {}
+  Grpc::AsyncClientFactoryPtr factoryForGrpcService(const envoy::config::core::v3::GrpcService&,
+                                                    Stats::Scope&, bool) override {
+    return std::make_unique<NiceMock<Grpc::MockAsyncClientFactory>>();
+  }
+};
 
-TEST(HttpExtAuthzConfigTest, CorrectProtoGoogleGrpc) {
-  std::string google_grpc_service_yaml = R"EOF(
-  transport_api_version: V3
-  grpc_service:
-    google_grpc:
-      target_uri: ext_authz_server
-      stat_prefix: google
-  failure_mode_allow: false
-  transport_api_version: V3
-  )EOF";
-  expectCorrectProtoGrpc(google_grpc_service_yaml);
-}
+class MultiThreadTest {
+public:
+  MultiThreadTest(size_t num_threads) : num_threads_(num_threads), api_(Api::createApiForTest()) {}
+  virtual ~MultiThreadTest() = default;
 
-TEST(HttpExtAuthzConfigTest, CorrectProtoEnvoyGrpc) {
-  std::string envoy_grpc_service_yaml = R"EOF(
-  transport_api_version: V3
-  grpc_service:
-    envoy_grpc:
-      cluster_name: ext_authz_server
-  failure_mode_allow: false
-  transport_api_version: V3
-  )EOF";
-  expectCorrectProtoGrpc(envoy_grpc_service_yaml);
-}
+  void postWorkToAllWorkers(std::function<void()> work) {
+    absl::Barrier start_barrier(num_threads_);
+    absl::BlockingCounter end_counter(num_threads_);
+
+    ASSERT(worker_dispatchers_.size() == num_threads_);
+    for (Event::DispatcherPtr& dispatcher : worker_dispatchers_) {
+      dispatcher->post([&, work]() {
+        start_barrier.Block();
+        work();
+        end_counter.DecrementCount();
+      });
+    }
+    // Wait for all workers to finish the job.
+    end_counter.Wait();
+  }
+
+  void postWorkToMain(std::function<void()> work) {
+    absl::BlockingCounter end_counter(1);
```

Comment:
`absl::Notification` probably makes more sense in this case.

Reply:
Thanks, removed.